flashinfer/deep_gemm/runtime.py

```diff
@@ -0,0 +1,275 @@
+    swizzle_mode: int,
+) -> cbd.CUtensorMap:
+    # `num_groups` is always applied into the outer dimensions
+    io_shapes = (shape_k, shape_n)[:: major_type.shape_direction()]
+    io_shapes = (io_shapes[0], io_shapes[1] * num_groups)
+
+    return make_tma_2d_desc(
+        t,
+        *io_shapes,
+        *(block_k, block_n)[:: major_type.shape_direction()],
+        outer_stride,
+        swizzle_mode,
+    )
+
+
+def make_tma_cd_desc(
+    major_type: MajorTypeCD,
+    t: torch.Tensor,
+    shape_m: int,
+    shape_n: int,
+    block_m: int,
+    block_n: int,
+    outer_stride: int,
+    num_groups: int,
+    swizzle_mode: int,
+) -> cbd.CUtensorMap:
+    assert major_type == MajorTypeCD.NMajor
+
+    # Swizzling requires the inner box dim to be less or equal than `kSwizzleCDMode`
+    # bytes, so `BLOCK_N * sizeof(T) / kSwizzleCDMode` TMA stores are required
+    layout_ad_m = 128
+    return make_tma_2d_desc(
+        t,
+        shape_n,
+        shape_m * num_groups,
+        block_n,
+        min(block_m, layout_ad_m),
+        outer_stride,
+        swizzle_mode,
+    )
+
+
+def make_tma_sf_desc(
+    major_type: MajorTypeAB,
+    t: torch.Tensor,
+    shape_mn: int,
+    shape_k: int,
+    block_mn: int,
+    block_k: int,
+    num_groups: int,
+    swizzle_mode: int,
+) -> cbd.CUtensorMap:
+    assert major_type == MajorTypeAB.MNMajor
+
+    # TODO: maybe swizzle SF as well
+    assert swizzle_mode == 0
+
+    # Make TMA aligned to 16 bytes
+    shape_mn = get_tma_aligned_size(shape_mn, t.element_size())
+    return make_tma_2d_desc(
+        t,
+        shape_mn,
+        ceil_div(shape_k, block_k * 4) * num_groups,
+        block_mn,
+        1,
+        shape_mn,
+        swizzle_mode,
+    )
+
+
+# Map some common Python types into C types
+pytypes_to_ctypes = {
+    True: "true",
+    False: "false",
+    torch.bfloat16: "cutlass::bfloat16_t",
+    torch.float: "float",
+}
+
+
+class Runtime:
+    def __init__(self, path: str) -> None:
+        self.path = path
+        self.lib = None
+        self.kernel = None
+        assert self.is_path_valid(self.path)
+
+    @staticmethod
+    def is_path_valid(path: str) -> bool:
+        # Exists and is a directory
+        if not os.path.exists(path) or not os.path.isdir(path):
+            return False
+
+        # Contains all necessary files
+        files = ["kernel.cubin"]
+        return all(os.path.exists(os.path.join(path, file)) for file in files)
+
+    @staticmethod
+    def generate(kwargs: Dict[str, Any]) -> str:
+        raise NotImplemented
+
+    @staticmethod
+    def launch(kernel: cbd.CUkernel, kwargs: Dict[str, Any]) -> cbd.CUresult:
+        raise NotImplemented
+
+    def __call__(self, **kwargs) -> cbd.CUresult:
+        # Load CUBIN
+        if self.kernel is None:
+
+            # Load CUBIN
+            path = bytes(os.path.join(self.path, "kernel.cubin"), "utf-8")
+            result, self.lib = cbd.cuLibraryLoadFromFile(path, [], [], 0, [], [], 0)
```

Comment:
Use [checkCudaErrors](https://github.com/flashinfer-ai/flashinfer/blob/728e8bb3ef4703ead30021f5e41c4bf7f22e81f8/flashinfer/cuda_utils.py#L35) for all cuda.bindings function calls.